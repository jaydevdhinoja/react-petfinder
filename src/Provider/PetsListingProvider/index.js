import React, { createContext, useContext, useState, useEffect } from 'react';
import { Client } from "@petfinder/petfinder-js";

const defaultState = {
	pets: [],
	searchTerm: ''
};

export const PetsListingContext = createContext(defaultState);

const client = new Client({apiKey: 'lZrY2wFLzhyEUefSPPn4JjEwIUBMP0gppdb1EEI9CXRRT629HL', secret: 'PgC6ma6wEX2Ll79BbODrGTX6IcKvwMnj7zVuAlOu'});

const PetsListingProvider = ({ children }) => {
	const InitialState = () => defaultState;

	const [ initialState, setInitialState ] = useState(InitialState());

	useEffect(
		() => {
			const loadPetsData = async () => {
				try {
					const animalType = {type: 'Dog' }
					const breed = initialState.searchTerm !== '' && initialState.searchTerm ? { breed: initialState.searchTerm } : {} 
					client.animal.search({...animalType, ...breed})
					.then(function (response) {
						let petfinderData = response.data;
						console.log(petfinderData)
						setInitialState({
							...initialState,
							pets: petfinderData.animals
						});
					})
					.catch(function (error) {
						console.log(error)
						setInitialState({
							...initialState, pets: []
						})
					});
				} catch (e) {
					console.error(e);
				}
			};
			loadPetsData();
		},
		[ initialState.searchTerm ]
	);

	return (
		<PetsListingContext.Provider
			value={{
				...initialState,
				setSearchTerm: (item) => {
					setInitialState({ ...initialState, searchTerm: item });
				}
			}}
		>
			{children}
		</PetsListingContext.Provider>
	);
};

export default PetsListingProvider;

export const usePetsListingContext = () => {
	return useContext(PetsListingContext);
};
