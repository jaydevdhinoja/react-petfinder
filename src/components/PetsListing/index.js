import React, { useState } from 'react';
import PetListItem from '../PetsListItem';
import { usePetsListingContext } from '../../Provider/PetsListingProvider';

const PetsListing = () => {
	const { pets, setSearchTerm } = usePetsListingContext();
	const [searchText, setSearchText ] = useState('');

	let listingItems = <div className="app_card-title">No items found!</div>;
	if (pets.length > 0) {
		listingItems = pets.map(animal => {
			return (
			  <PetListItem
				key={animal.id}
				id={animal.id}
				name={animal.name}
				photos={animal.photos}
				primaryBreed={animal.breeds.primary}
				secondaryBreed={animal.breeds.secondary}
				description={animal.description}
			  />
			);
		  })
	}

	const onSearch = (e) => {
		e.preventDefault()
		setSearchTerm(searchText)
	}

	return (
		<form onSubmit={onSearch}>
			<div className="app_container">
				<div>
					Pet Finder
				</div>
				<div className="app_center_heading">
					<div className="app_center_title">
						Search
					</div>
					<div className="app_center_wrap">
						<input type="text" className="app_center_searchTerm" placeholder="What are you looking for?" value={searchText} onChange={e => setSearchText(e.target.value)}/>
					</div>
				</div>
				<div className="app_box">{listingItems}</div>
			</div>
		</form>
	);
};

export default PetsListing;
