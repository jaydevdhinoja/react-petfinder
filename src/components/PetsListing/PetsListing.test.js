import '@testing-library/jest-dom/extend-expect';
import React from 'react';
import PetsListing from '.';
import { PetsListingContext } from '../../Provider/PetsListingProvider';
import { render, cleanup } from '@testing-library/react';
import { createMemoryHistory } from 'history'
import { Router } from 'react-router-dom'
import { MemoryRouter } from 'react-router-dom'

afterEach(cleanup);

const InitialState = {
	pets: [
		{
			id: 47782140,
			organization_id: "FL850",
			url: "https://www.petfinder.com/dog/fred-47782140/fl/miami-beach/dachshund-rescue-south-florida-fl850/?referrer_id=b5f7363e-4d07-4756-ab82-d3a8060feccd",
			type: "Dog",
			species: "Dog",
			breeds: {primary: "Dachshund", secondary: null, mixed: false, unknown: false},
			colors: {primary: "Red / Chestnut / Orange", secondary: null, tertiary: null},
			age: "Adult",
			gender: "Male",
			size: "Small",
			name: "Fred",
			coat: "Short",
			description: "Can you say &quot;Personality? This is Fred and he has is in loads!  He is a handsome 9 year old...",
			photos: [
				{
					small: "https://dl5zpyw5k3jeb.cloudfront.net/photos/pets/47782140/5/?bust=1586432382&width=100",
					medium: "https://dl5zpyw5k3jeb.cloudfront.net/photos/pets/47782140/5/?bust=1586432382&width=300",
					large: "https://dl5zpyw5k3jeb.cloudfront.net/photos/pets/47782140/5/?bust=1586432382&width=600",
					full: "https://dl5zpyw5k3jeb.cloudfront.net/photos/pets/47782140/5/?bust=1586432382"
				}
			]
		}
	],
	searchTerm: ''
};

function renderWithContext(node, { value, ...options }) {
	return render(<PetsListingContext.Provider value={value}>{node}</PetsListingContext.Provider>, options);
}

describe('Pet Listing', () => {
	test('renders component', () => {
		const { container } = renderWithContext(<PetsListing />, { value: InitialState, wrapper: MemoryRouter });
		expect(container).toMatchSnapshot();
	});

	test('renders component no pets', () => {
		const { getByText } = renderWithContext(<PetsListing />, {
			value: { pets:[], searchTerm:'' },
			wrapper: MemoryRouter
		});

		expect(getByText(/No items found/i)).toBeInTheDocument();
	});

	test('renders pets listing', () => {
		const { getByText } = renderWithContext(<PetsListing />, { value: InitialState, wrapper: MemoryRouter });
		expect(getByText('Fred')).toBeInTheDocument();
	});

});
