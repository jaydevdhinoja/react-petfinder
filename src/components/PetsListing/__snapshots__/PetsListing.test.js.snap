// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pet Listing renders component 1`] = `
<div>
  <form>
    <div
      class="app_container"
    >
      <div>
        Pet Finder
      </div>
      <div
        class="app_center_heading"
      >
        <div
          class="app_center_title"
        >
          Search
        </div>
        <div
          class="app_center_wrap"
        >
          <input
            class="app_center_searchTerm"
            placeholder="What are you looking for?"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="app_box"
      >
        <div
          class="app_card"
        >
          <div
            class="app_card-image"
            style="background-image: url(https://dl5zpyw5k3jeb.cloudfront.net/photos/pets/47782140/5/?width=200);"
          >
            <img
              alt="Fred"
              src="https://dl5zpyw5k3jeb.cloudfront.net/photos/pets/47782140/5/?width=200"
              style="display: none;"
            />
          </div>
          <div
            class="app_card-details"
          >
            <h3>
              Fred
            </h3>
            <p>
              Primary Breed: Dachshund
            </p>
            <p>
              Secondary Breed: 
            </p>
          </div>
        </div>
      </div>
    </div>
  </form>
</div>
`;
